go/mysql/conn.go

```diff
@@ -758,383 +758,430 @@ func (c *Conn) writeEOFPacket(flags uint16, warnings uint16) error {
+		if err == nil || err == io.EOF {
+			err = NewSQLErrorFromError(errors.New("unexpected: query ended without no results and no error"))
 		}
+		if werr := c.writeErrorPacketFromError(err); werr != nil {
+			// If we can't even write the error, we're done.
+			log.Errorf("Error writing query error to %s: %v", c, werr)
+			return false
+		}
+		return true
+	}
+	if err != nil {
+		// We can't send an error in the middle of a stream.
+		// All we can do is abort the send, which will cause a 2013.
+		log.Errorf("Error in the middle of a stream to %s: %v", c, err)
+		return false
+	}
 
-		if err := c.writeOKPacket(0, 0, c.StatusFlags, 0); err != nil {
-			log.Error("Error writing ComStmtReset OK packet to client %v: %v", c.ConnectionID, err)
-			return err
+	// Send the end packet only sendFinished is false (results were streamed).
+	// In this case the affectedRows and lastInsertID are always 0 since it
+	// was a read operation.
+	if !sendFinished {
+		if err := c.writeEndResult(false, 0, 0, handler.WarningCount(c)); err != nil {
+			log.Errorf("Error writing result to %s: %v", c, err)
+			return false
 		}
+	}
 
-	case ComResetConnection:
-		// Clean up and reset the connection
-		c.recycleReadPacket()
-		handler.ComResetConnection(c)
-		// Reset prepared statements
-		c.PrepareData = make(map[uint32]*PrepareData)
-		err = c.writeOKPacket(0, 0, 0, 0)
-		if err != nil {
-			c.writeErrorPacketFromError(err)
+	timings.Record(queryTimingKey, queryStart)
+	return true
+}
+
+func (c *Conn) splitQueries(query string) ([]string, error) {
+	if c.Capabilities&CapabilityClientMultiStatements == 0 {
+		return []string{query}, nil
+	}
+
+	queries, err := sqlparser.SplitStatementToPieces(query)
+	if err != nil {
+		log.Errorf("Conn %v: Error splitting query: %v", c, err)
+		c.startWriterBuffering()
+
+		if werr := c.writeErrorPacketFromError(err); werr != nil {
+			// If we can't even write the error, we're done.
+			log.Errorf("Conn %v: Error writing query error: %v", c, werr)
+			return nil, werr
 		}
 
-	default:
-		log.Errorf("Got unhandled packet (default) from %s, returning error: %v", c, data)
-		c.recycleReadPacket()
-		if err := c.writeErrorPacket(ERUnknownComError, SSUnknownComError, "command handling not implemented yet: %v", data[0]); err != nil {
-			log.Errorf("Error writing error packet to %s: %s", c, err)
-			return err
+		if err := c.endWriterBuffering(); err != nil {
+			log.Errorf("conn %v: flush() failed: %v", c.ID(), err)
 		}
+		return nil, err
 	}
 
-	return nil
+	return queries, nil
+}
+
+func (c *Conn) execBufferedQuery(handler Handler, sql string, more bool) execResult {
+	c.startWriterBuffering()
+	defer func() {
+		if err := c.endWriterBuffering(); err != nil {
+			log.Errorf("conn %v: flush() failed: %v", c.ID(), err)
+		}
+	}()
+
+	return c.execQuery(sql, handler, more)
 }
 
-func (c *Conn) execQuery(query string, handler Handler, more bool) error {
-	fieldSent := false
+type execResult int8
+
+const (
+	success execResult = iota
```

Comment:
Minor naming nit -- the reference to `success` below kind confused me... maybe `execSuccess` to be clearer and more consistent with the others?